File: shoring/blackjackEngine/src/test/java/com/panopset/tests/blackjackEngine/CountingSystemsTest.java
package com.panopset.tests.blackjackEngine;


import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotEquals;
import org.junit.jupiter.api.Test;

import com.panopset.blackjackEngine.BlackjackCard;
import com.panopset.blackjackEngine.BlackjackConfigDefault;
import com.panopset.blackjackEngine.Card;
import com.panopset.blackjackEngine.CardCounterImpl;
import com.panopset.blackjackEngine.CardDefinition;
import com.panopset.blackjackEngine.CountingSystem;

public class CountingSystemsTest {

  @Test
  void test() throws Exception {
    CardCounterImpl.updateByIndex(3);
    testConfig();
    CardCounterImpl.setConfig(BlackjackConfigDefault.INSTANCE);
    CardCounterImpl.updateByIndex(3);
    testConfig();
  }
  
  private void testConfig() throws Exception {
    CountingSystem dft = CardCounterImpl.getSystem();
    CardCounterImpl.resetCount();
    assertEquals(8, CardCounterImpl.getKeyNames().size());    
    assertEquals(0, CardCounterImpl.getCount());
    new BlackjackCard(new Card(CardDefinition.FIVE_OF_CLUBS));
    assertEquals(1, CardCounterImpl.getCount());
    new BlackjackCard( new Card(CardDefinition.FOUR_OF_HEARTS));
    assertEquals(2, CardCounterImpl.getCount());
    new BlackjackCard(new Card(CardDefinition.EIGHT_OF_HEARTS));
    assertEquals(2, CardCounterImpl.getCount());
    new BlackjackCard(new Card(CardDefinition.ACE_OF_HEARTS));
    assertEquals(1, CardCounterImpl.getCount());
    new BlackjackCard(new Card(CardDefinition.TEN_OF_SPADES));
    assertEquals(0, CardCounterImpl.getCount());
    new BlackjackCard(new Card(CardDefinition.THREE_OF_SPADES));
    assertEquals(1, CardCounterImpl.getCount());
    new BlackjackCard(new Card(CardDefinition.TWO_OF_HEARTS));
    assertEquals(2, CardCounterImpl.getCount());
    CardCounterImpl.updateByIndex(0);
    assertEquals("None", CardCounterImpl.getSystem().getName());
    CardCounterImpl.updateByIndex(1);
    assertEquals("Wizard Ace/5", CardCounterImpl.getSystem().getName());
    CardCounterImpl.updateByIndex(2);
    CountingSystem ko = CardCounterImpl.getSystem();
    assertEquals("KO", ko.getName());
    CardCounterImpl.updateByIndex(3);
    CountingSystem hiLo = CardCounterImpl.getSystem();
    assertEquals(hiLo.hashCode(), dft.hashCode());
    assertEquals(hiLo, dft);
    assertNotEquals(hiLo, ko);
    assertEquals("Hi-Lo", hiLo.getName());
    CardCounterImpl.updateByIndex(4);
    assertEquals("Hi-Opt I", CardCounterImpl.getSystem().getName());
    CardCounterImpl.updateByIndex(5);
    assertEquals("Hi-Opt II", CardCounterImpl.getSystem().getName());
    CardCounterImpl.updateByIndex(6);
    assertEquals("Zen Count", CardCounterImpl.getSystem().getName());
    CardCounterImpl.updateByIndex(7);
    assertEquals("Omega II", CardCounterImpl.getSystem().getName());
    assertEquals(7, CardCounterImpl.INSTANCE.getCountingSystems().getKeyNamePosition("Omega II"));
    assertEquals(3, CardCounterImpl.INSTANCE.getCountingSystems().getKeyNamePosition("foo"));
    assertEquals("Hi-Lo", CardCounterImpl.INSTANCE.getCountingSystems().findSelectionKey(-1));
  }

}

File: code/shoring/blackjackEngine/src/test/kotlin/com/panopset/tests/blackjackEngine/CountingSystemsTest.kt
package com.panopset.tests.blackjackEngine

import com.panopset.blackjackEngine.*
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

class CountingSystemsTest {
    private val countingSystems = CountingSystems(BlackjackConfigDefault())

    @Test
    @Throws(Exception::class)
    fun test() {
        countingSystems.updateByIndex(3)
        testConfig()
        countingSystems.updateByIndex(3)
        testConfig()
    }

    @Throws(Exception::class)
    private fun testConfig() {
        val dft = countingSystems.findSelected()
        countingSystems.resetCount()
        Assertions.assertEquals(8, countingSystems.keyNames.size)
        Assertions.assertEquals(0, countingSystems.getTrueCount())
        countingSystems.count(BlackjackCard(Card(CardDefinition.FIVE_OF_CLUBS)))
        Assertions.assertEquals(1, countingSystems.getTrueCount())
        countingSystems.count(BlackjackCard(Card(CardDefinition.FOUR_OF_HEARTS)))
        Assertions.assertEquals(2, countingSystems.getTrueCount())
        countingSystems.count(BlackjackCard(Card(CardDefinition.EIGHT_OF_HEARTS)))
        Assertions.assertEquals(2, countingSystems.getTrueCount())
        countingSystems.count(BlackjackCard(Card(CardDefinition.ACE_OF_HEARTS)))
        Assertions.assertEquals(1, countingSystems.getTrueCount())
        countingSystems.count(BlackjackCard(Card(CardDefinition.TEN_OF_SPADES)))
        Assertions.assertEquals(0, countingSystems.getTrueCount())
        countingSystems.count(BlackjackCard(Card(CardDefinition.THREE_OF_SPADES)))
        Assertions.assertEquals(1, countingSystems.getTrueCount())
        countingSystems.count(BlackjackCard(Card(CardDefinition.TWO_OF_HEARTS)))
        Assertions.assertEquals(2, countingSystems.getTrueCount())
        countingSystems.updateByIndex(0)
        Assertions.assertEquals("None", countingSystems.findSelected().name)
        countingSystems.updateByIndex(1)
        Assertions.assertEquals("Wizard Ace/5", countingSystems.findSelected().name)
        countingSystems.updateByIndex(2)
        val ko = countingSystems.findSelected()
        Assertions.assertEquals("KO", ko.name)
        countingSystems.updateByIndex(3)
        val hiLo = countingSystems.findSelected()
        Assertions.assertEquals(hiLo.hashCode(), dft.hashCode())
        Assertions.assertEquals(hiLo, dft)
        Assertions.assertNotEquals(hiLo, ko)
        Assertions.assertEquals("Hi-Lo", hiLo.name)
        countingSystems.updateByIndex(4)
        Assertions.assertEquals("Hi-Opt I", countingSystems.findSelected().name)
        countingSystems.updateByIndex(5)
        Assertions.assertEquals("Hi-Opt II", countingSystems.findSelected().name)
        countingSystems.updateByIndex(6)
        Assertions.assertEquals("Zen Count", countingSystems.findSelected().name)
        countingSystems.updateByIndex(7)
        Assertions.assertEquals("Omega II", countingSystems.findSelected().name)
        Assertions.assertEquals(7, countingSystems.getKeyNamePosition("Omega II"))
        Assertions.assertEquals(3, countingSystems.getKeyNamePosition("foo"))
        Assertions.assertEquals("Hi-Lo", countingSystems.findSelectionKey(-1))
    }
}
